### ISSUE TITLE
Feature request: photo/thumbnail browser

### ISSUE THREAD

**Issue by tridian-tn** [yyoon/Journaley#104]
This is something I wish Day One had, and now by extension, Journaley too: I journal in conjunction with my photos, so I will snap away with my phone and then take the time later to go through my day and write about it with the help of the photos, attaching the photo to the journal entry as I go. What this has also meant is that when I look at the journal entry a few weeks later, I wonder what else may've happened that day. So I then look in my photos folder (on Dropbox, but regardless: it's just a folder) to see more about the day.

For me personally, it would be great to have a (maybe fly out) panel in Journaley that (given an indexed folder in the settings/preferences) dynamically shows thumbnails from the same period as the journal entry. So e.g. if there's one entry that day, it shows thumbnails for the day. If there is more than one journal entry, it shows the photos since the last entry and before the next one. 

In a nutshell it provides context to the journal entry, and should inspire to write more about forgotten parts of the day that were impulsively captured in a photo.

**Comment by sguergachi** [comment]
Let me see if I get what you're saying. You take multiple photo's of your day that you intend to journal about later on. You then go through your photos for that day and write a single entry with one of the photo's in mind (or do you write an entry for each photo?). And you sometimes go back to these entries and wonder about the other pictures that you _didn't_ put into an entry? So instead of having them accessible in the app you go through your photos folder and manually look for the _other_ photos you took that day. Now you're asking if those _other_ photos can be accessible from within the entry?

Not sure if what I understood what you're asking, but hopefully I got some of it.

Here are questions I have:
- Do you have an entry for each photo you took that day?
- If so are you asking for an easier way to see all the photo's of that day from _any_ entry, rather than browsing to the other entries for that day?
- Or are you asking if you can attach _multiple_ photos to a single entry?

**Comment by tridian-tn** [comment]
Thanks for replying. There are two parts to the process here:
- Journalling: I happily browse through photos and add journal entries if any of those photos inspire me to write about what happened. The current limitation of one photo per entry is absolutely fine here, and to be honest the current workflow of 'attach a photo, use system file browser to find it' is perfectly serviceable.
- Reading back: this is the part that I feel would be enriched by a passive photos panel. I often look at old entries and then get the urge to dig out the photos of that day for a bit of context. Having them in eyeshot (albeit as a smaller thumbnail grid, say) would make that process (essentially reminiscing) easier and because of its emotive aspect, probably quite addictive.

To specifically answer your Qs:
- "Do you have an entry for each photo you took that day?" No, but I do pick a few
- "If so are you asking for an easier way to see all the photo's of that day from any entry, rather than browsing to the other entries for that day?" Yes, absolutely
- "Or are you asking if you can attach multiple photos to a single entry?" No

Thinking ahead, having a 'right click on thumbnail, context menu pops up with Create Entry With Photo' feature would be a logical step from a dynamic thumbnail panel, but this would only be of secondary benefit to me.

**Comment by sguergachi** [comment]
I'm going to write out my thought process just to have all the info out there.

Ok, so those additional - let's call them "contextual photos" - how would they be programmatically accessed? I think we would be depending on date metadata attached to the picture file, so we have to accept that partial or no photo's may be visible, and that users might not understand _why_ certain pictures aren't showing up because the necessary metadata is not accessible. 

Considering this isn't part of Day One's data model, it would be a divergence we would have to properly figure out. With Day One, each entry photo has its own unique identifier to link to its respective entry. 

So what you are saying is that each entry has one "main" photo and it would additionally look through a directory for all contextual photos that match the same _day_ as the _main_ photo, since you might be creating the entry on a day different from when the photos were actually taken.

I think I got what you're asking then. Sounds doable. We just have to figure out what @yyoon thinks and whether this is something worth making.

Regardless, I'm not sure this is a feature we're going to implement just yet. Our main focus right now is to at the very least be in parity, feature wise, with Day One, and be able to properly access (and perhaps edit) all the metadata Day One entries provide, like Weather and Location for instance. We wouldn't be at a point of making diverging features from Day One just yet (not until after 2.5 I would say). 

This is still a good idea. I've been thinking a lot about contextualizing entries, #56 for example is one feature we both think we would use as people who spend a lot of their time on the computer. Definately something we can come back to, and now that I think we've fleshed out some specifics, it will be easier to look back at this and implement.

**Comment by yyoon** [comment]
@pellenys First of all, thanks for coming here and leaving a feature request. This could be a good way to overcome the limitation of Day One that an entry can only have one photo attached to it. I cannot think of a good UI for this feature on top of my head, but @sguergachi might have a better idea. In terms of implementation, it wouldn't be too hard, if the user keeps all the photos in one folder, and specifies that particular photo folder in the settings UI (could default to My Pictures). Or, we could make the app even search for photos in all the sub-folders of the specified folder.

I agree that this is something we probably wouldn't do right away. The use case you described kind of makes sense, but I'm not sure how many users would have the same kind of desire.

**Comment by sguergachi** [comment]
We should also probably wait for Day One 2.0 before making any decision. Seems like multiple-photo's per entry is one of many features they plan on adding according to this: http://dayoneapp.com/sync/

Though it's not clear whether Dropbox synced entries will get this features.

**Comment by tridian-tn** [comment]
Thanks for considering this. I appreciate that this is a step beyond Day One, but if it's in the wishlist for the future, that's great.

re "each entry has one "main" photo and it would additionally look through a directory for all contextual photos that match the same day as the main photo, since you might be creating the entry on a day different from when the photos were actually taken", I was thinking that it would be based on the editable journal entry datestamp, rather than requiring an entry to have a photo. It would also be limited to the entry view UI rather than the entry edit UI, removing any need to monitor the edited date and refresh accordingly,

re metadata vs date, I'm coming from a scenario where the photos always have the date in the metadata from the camera or 'phone, so you'd rely on that rather than (say) last-modified, which would also handily filter out misc images without metadata such as screenshots and wallpaper. I acknowledge that this doesn't help if this metadata is wrongly datestamped (e.g. my mother in law's photos that always come up as 1/1/2000 because she never bothers to reset the date on her camera), but I suppose you've got to stop somewhere.

re photo location, I was imagining a recursive search through a single folder, yes.....although I don't have any idea what the performance of that would be if it wasn't helped with some form of cache....which brings a new level of complexity unless the Windows Search Index already caches photo metadata? No idea - haven't done Windows file-system level coding for some time.

There's a definite appeal to features such as this that don't affect the stored data, but yes - it will be interesting to see what Day One come up with regarding multiple images.

Again, thanks for considering my suggestion!

**Comment by sguergachi** [comment]
I do wonder about the performance implications now too. I'm guessing caching will happen at the creation of the entry, but if we are saying that we need to keep an updated cache for possibly new pictures, then we would likely have to update the cache every time the entry loads. @yyoon might be able to test this, if and when we decide to do this.
